File: src/assign-cgroups.py
#!/usr/bin/python3
"""
Automatically assign a dedicated systemd scope to the GUI applications
launched in the same cgroup as the compositor. This could be helpful for
implementing cgroup-based resource management and would be necessary when
`systemd-oomd` is in use.

Limitations: The script is using i3ipc window:new event to detect application
launches and would fail to detect background apps or special surfaces.
Therefore it's recommended to supplement the script with use of systemd user
services for such background apps.

Dependencies: dbus-next, i3ipc, psutil, tenacity, python-xlib
"""
import argparse
import asyncio
import logging
import re
import socket
import struct
import sys
from functools import lru_cache
from typing import Optional

from dbus_next import Variant
from dbus_next.aio import MessageBus
from dbus_next.errors import DBusError
from i3ipc import Event
from i3ipc.aio import Con, Connection
from psutil import Process
from tenacity import retry, retry_if_exception_type, stop_after_attempt

if sys.version_info[:2] >= (3, 9):
    from collections.abc import Callable
else:
    from typing import Callable


LOG = logging.getLogger("assign-cgroups")
SD_BUS_NAME = "org.freedesktop.systemd1"
SD_OBJECT_PATH = "/org/freedesktop/systemd1"
SD_SLICE_FORMAT = "app-{app_id}.slice"
SD_UNIT_FORMAT = "app-{app_id}-{unique}.scope"
# Ids of known launcher applications that are not special surfaces. When the app is
# started using one of those, it should be moved to a new cgroup.
# Launcher should only be listed here if it creates cgroup of its own.
LAUNCHER_APPS = ["nwgbar", "nwgdmenu", "nwggrid", "onagre"]

SD_UNIT_ESCAPE_RE = re.compile(r"[^\w:.\\]", re.ASCII)


def escape_app_id(app_id: str) -> str:
    """Escape app_id for systemd APIs.

    The "unit prefix" must consist of one or more valid characters (ASCII letters,
    digits, ":", "-", "_", ".", and "\"). The total length of the unit name including
    the suffix must not exceed 256 characters. [systemd.unit(5)]

    We also want to escape "-" to avoid creating extra slices.
    """

    def repl(match):
        return "".join([f"\\x{x:02x}" for x in match.group().encode()])

    return SD_UNIT_ESCAPE_RE.sub(repl, app_id)


LAUNCHER_APP_CGROUPS = [
    SD_SLICE_FORMAT.format(app_id=escape_app_id(app)) for app in LAUNCHER_APPS
]


def get_cgroup(pid: int) -> Optional[str]:
    """
    Get cgroup identifier for the process specified by pid.
    Assumes cgroups v2 unified hierarchy.
    """
    try:
        with open(f"/proc/{pid}/cgroup", "r") as file:
            cgroup = file.read()
        return cgroup.strip().split(":")[-1]
    except OSError:
        LOG.exception("Error geting cgroup info")
    return None


def get_pid_by_socket(sockpath: str) -> int:
    """
    getsockopt (..., SO_PEERCRED, ...) returns the following structure
    struct ucred
    {
      pid_t pid; /* s32: PID of sending process.  */
      uid_t uid; /* u32: UID of sending process.  */
      gid_t gid; /* u32: GID of sending process.  */
    };
    See also: socket(7), unix(7)
    """
    with socket.socket(socket.AF_UNIX, socket.SOCK_STREAM) as sock:
        sock.connect(sockpath)
        ucred = sock.getsockopt(
            socket.SOL_SOCKET, socket.SO_PEERCRED, struct.calcsize("iII")
        )
    pid, _, _ = struct.unpack("iII", ucred)
    return pid


def create_x11_pid_getter() -> Callable[[int], int]:
    """Create fallback X11 PID getter.

    Sway 1.6.1/wlroots 0.14 can use XRes to get the PID for Xwayland apps from
    the server and won't ever reach that. The fallback is preserved for
    compatibility with i3 and earlier versions of Sway.
    """
    # pylint: disable=import-outside-toplevel
    # Defer Xlib import until we really need it.
    from Xlib import X
    from Xlib.display import Display

    try:
        # requires python-xlib >= 0.30
        from Xlib.ext import res as XRes
    except ImportError:
        XRes = None

    display = Display()

    def get_net_wm_pid(wid: int) -> int:
        """Get PID from _NET_WM_PID property of X11 window"""
        window = display.create_resource_object("window", wid)
        net_wm_pid = display.get_atom("_NET_WM_PID")
        pid = window.get_full_property(net_wm_pid, X.AnyPropertyType)

        if pid is None:
            raise RuntimeError("Failed to get PID from _NET_WM_PID")
        return int(pid.value.tolist()[0])

    def get_xres_client_id(wid: int) -> int:
        """Get PID from X server via X-Resource extension"""
        res = display.res_query_client_ids(
            [{"client": wid, "mask": XRes.LocalClientPIDMask}]
        )
        for cid in res.ids:
            if cid.spec.client > 0 and cid.spec.mask == XRes.LocalClientPIDMask:
                for value in cid.value:
                    return value
        raise RuntimeError("Failed to get PID via X-Resource extension")

    if XRes is None or display.query_extension(XRes.extname) is None:
        LOG.warning(
            "X-Resource extension is not supported. "
            "Process identification for X11 applications will be less reliable."
        )
        return get_net_wm_pid

    ver = display.res_query_version()
    LOG.info(
        "X-Resource version %d.%d",
        ver.server_major,
        ver.server_minor,
    )
    if (ver.server_major, ver.server_minor) < (1, 2):
        return get_net_wm_pid

    return get_xres_client_id


class CGroupHandler:
    """Main logic: handle i3/sway IPC events and start systemd transient units."""

    def __init__(self, bus: MessageBus, conn: Connection):
        self._bus = bus
        self._conn = conn

    @property
    @lru_cache(maxsize=1)
    def get_x11_window_pid(self) -> Optional[Callable[[int], int]]:
        """On-demand initialization of X11 PID getter"""
        try:
            return create_x11_pid_getter()
        # pylint: disable=broad-except
        except Exception as exc:
            LOG.warning("Failed to create X11 PID getter: %s", exc)
            return None

    async def connect(self):
        """asynchronous initialization code"""
        # pylint: disable=attribute-defined-outside-init
        introspection = await self._bus.introspect(SD_BUS_NAME, SD_OBJECT_PATH)
        self._sd_proxy = self._bus.get_proxy_object(
            SD_BUS_NAME, SD_OBJECT_PATH, introspection
        )
        self._sd_manager = self._sd_proxy.get_interface(f"{SD_BUS_NAME}.Manager")

        self._compositor_pid = get_pid_by_socket(self._conn.socket_path)
        self._compositor_cgroup = get_cgroup(self._compositor_pid)
        assert self._compositor_cgroup is not None
        LOG.info("compositor:%s %s", self._compositor_pid, self._compositor_cgroup)

        self._conn.on(Event.WINDOW_NEW, self._on_new_window)
        return self

    def get_pid(self, con: Con) -> Optional[int]:
        """Get PID from IPC response (sway), X-Resource or _NET_WM_PID (i3)"""
        if isinstance(con.pid, int) and con.pid > 0:
            return con.pid

        if con.window is not None and self.get_x11_window_pid is not None:
            return self.get_x11_window_pid(con.window)

        return None

    def cgroup_change_needed(self, cgroup: Optional[str]) -> bool:
        """Check criteria for assigning current app into an isolated cgroup"""
        if cgroup is None:
            return False
        for launcher in LAUNCHER_APP_CGROUPS:
            if launcher in cgroup:
                return True
        return cgroup == self._compositor_cgroup

    @retry(
        reraise=True,
        retry=retry_if_exception_type(DBusError),
        stop=stop_after_attempt(3),
    )
    async def assign_scope(self, app_id: str, proc: Process):
        """
        Assign process (and all unassigned children) to the
        app-{app_id}.slice/app{app_id}-{pid}.scope cgroup
        """
        app_id = escape_app_id(app_id)
        sd_slice = SD_SLICE_FORMAT.format(app_id=app_id)
        sd_unit = SD_UNIT_FORMAT.format(app_id=app_id, unique=proc.pid)
        # Collect child processes as systemd assigns a scope only to explicitly
        # specified PIDs.
        # There's a risk of race as the child processes may exit by the time dbus call
        # reaches systemd, hence the @retry decorator is applied to the method.
        pids = [proc.pid] + [
            x.pid
            for x in proc.children(recursive=True)
            if self.cgroup_change_needed(get_cgroup(x.pid))
        ]

        await self._sd_manager.call_start_transient_unit(
            sd_unit,
            "fail",
            [["PIDs", Variant("au", pids)], ["Slice", Variant("s", sd_slice)]],
            [],
        )
        LOG.debug(
            "window %s successfully assigned to cgroup %s/%s", app_id, sd_slice, sd_unit
        )

    async def _on_new_window(self, _: Connection, event: Event):
        """window:new IPC event handler"""
        con = event.container
        app_id = con.app_id if con.app_id else con.window_class
        try:
            pid = self.get_pid(con)
            if pid is None:
                LOG.warning("Failed to get pid for %s", app_id)
                return
            proc = Process(pid)
            cgroup = get_cgroup(proc.pid)
            # some X11 apps don't set WM_CLASS. fallback to process name
            if app_id is None:
                app_id = proc.name()
            LOG.debug("window %s(%s) cgroup %s", app_id, proc.pid, cgroup)
            if self.cgroup_change_needed(cgroup):
                await self.assign_scope(app_id, proc)
        # pylint: disable=broad-except
        except Exception as exc:
            LOG.error("Failed to modify cgroup for %s: %s", app_id, exc)


async def main():
    """Async entrypoint"""
    try:
        bus = await MessageBus().connect()
        conn = await Connection(auto_reconnect=False).connect()
        await CGroupHandler(bus, conn).connect()
        await conn.main()
    except DBusError as exc:
        LOG.error("DBus connection error: %s", exc)
    except (ConnectionError, EOFError) as exc:
        LOG.error("Sway IPC connection error: %s", exc)


if __name__ == "__main__":
    parser = argparse.ArgumentParser(
        description="Assign CGroups to apps in compositors with i3 IPC protocol support"
    )
    parser.add_argument(
        "-l",
        "--loglevel",
        choices=["critical", "error", "warning", "info", "debug"],
        default="info",
        dest="loglevel",
        help="set logging level",
    )
    args = parser.parse_args()
    logging.basicConfig(level=args.loglevel.upper())
    asyncio.run(main())

File: src/assign-cgroups.rs
//! Automatically assign a dedicated systemd scope to the GUI applications
//! launched in the same cgroup as the compositor. This could be helpful for
//! implementing cgroup-based resource management and would be necessary when
//! `systemd-oomd` is in use.
//!
//! Limitations: The script is using i3ipc window:new event to detect application
//! launches and would fail to detect background apps or special surfaces.
//! Therefore it's recommended to supplement the script with use of systemd user
//! services for such background apps.

use clap::{ArgEnum, Parser};
use dbus::{
    arg::{RefArg, Variant},
    blocking::Connection as DbusConnection,
};
use futures_util::StreamExt;
use log::{debug, error, info, warn};
use std::sync::Arc;
use std::{
    fs, io,
    os::unix::{io::AsRawFd, net::UnixStream},
    time::Duration,
};
use swayipc_async::{Connection as SwayConnection, Event, EventType, WindowChange, WindowEvent};
use tokio::sync::Mutex;

const SD_BUS_NAME: &str = "org.freedesktop.systemd1";
const SD_OBJECT_PATH: &str = "/org/freedesktop/systemd1";
const SD_MANAGER_IFACE: &str = "org.freedesktop.systemd1.Manager";

const SD_SLICE_FORMAT: &str = "app-{app_id}.slice";
const SD_UNIT_FORMAT: &str = "app-{app_id}-{unique}.scope";

/// Launcher apps that create their own cgroup; windows spawned from them
/// should still be moved to a new scope.
const LAUNCHER_APPS: &[&str] = &["nwgbar", "nwgdmenu", "nwggrid", "onagre"];

#[repr(C)]
struct UCred {
    pid: i32,
    uid: u32,
    gid: u32,
}

fn get_pid_by_socket(path: &str) -> io::Result<u32> {
    let sock = UnixStream::connect(path)?;
    let mut ucred = UCred {
        pid: 0,
        uid: 0,
        gid: 0,
    };
    let mut len = std::mem::size_of::<UCred>() as libc::socklen_t;
    let ret = unsafe {
        libc::getsockopt(
            sock.as_raw_fd(),
            libc::SOL_SOCKET,
            libc::SO_PEERCRED,
            &mut ucred as *mut UCred as *mut libc::c_void,
            &mut len,
        )
    };
    if ret != 0 {
        return Err(io::Error::last_os_error());
    }
    Ok(ucred.pid as u32)
}

// cgroup helpers

fn get_cgroup(pid: u32) -> Option<String> {
    let text = fs::read_to_string(format!("/proc/{pid}/cgroup")).ok()?;
    // cgroupv2: single line "0::<path>"; v1 has multiple lines — take last field
    text.lines()
        .next()
        .and_then(|l| l.split(':').last())
        .map(|s| s.trim().to_string())
}

/// Walk /proc/<pid>/task/<tid>/children recursively.
fn collect_child_pids(root: u32) -> Vec<u32> {
    let mut result = Vec::new();
    let mut stack = vec![root];
    while let Some(pid) = stack.pop() {
        let task_dir = format!("/proc/{pid}/task");
        let Ok(entries) = fs::read_dir(&task_dir) else {
            continue;
        };
        for entry in entries.flatten() {
            let children_path = entry.path().join("children");
            let Ok(data) = fs::read_to_string(&children_path) else {
                continue;
            };
            for token in data.split_whitespace() {
                if let Ok(child) = token.parse::<u32>() {
                    result.push(child);
                    stack.push(child);
                }
            }
        }
    }
    result
}

fn escape_app_id(app_id: &str) -> String {
    let mut out = String::with_capacity(app_id.len() * 2);
    for b in app_id.bytes() {
        match b {
            b'a'..=b'z' | b'A'..=b'Z' | b'0'..=b'9' | b':' | b'.' | b'\\' | b'_' => {
                out.push(b as char);
            }
            other => {
                out.push_str(&format!("\\x{:02x}", other));
            }
        }
    }
    out
}

struct CGroupHandler {
    compositor_cgroup: String,
    launcher_cgroups: Vec<String>,
    dbus_conn: Arc<Mutex<DbusConnection>>,
}

impl CGroupHandler {
    fn new(compositor_cgroup: String, dbus_conn: Arc<Mutex<DbusConnection>>) -> Self {
        let launcher_cgroups = LAUNCHER_APPS
            .iter()
            .map(|app| SD_SLICE_FORMAT.replace("{app_id}", &escape_app_id(app)))
            .collect();
        Self {
            compositor_cgroup,
            launcher_cgroups,
            dbus_conn,
        }
    }

    fn cgroup_change_needed(&self, cgroup: Option<&str>) -> bool {
        let Some(cg) = cgroup else { return false };
        if cg == self.compositor_cgroup {
            return true;
        }
        for launcher in &self.launcher_cgroups {
            if cg.contains(launcher.as_str()) {
                return true;
            }
        }
        false
    }

    /// Assign `pid` and unassigned children to a transient systemd scope.
    /// Retries up to 3 times on D-Bus error
    async fn assign_scope(
        &self,
        app_id: &str,
        pid: u32,
    ) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
        let escaped = escape_app_id(app_id);
        let sd_slice = SD_SLICE_FORMAT.replace("{app_id}", &escaped);
        let sd_unit = SD_UNIT_FORMAT
            .replace("{app_id}", &escaped)
            .replace("{unique}", &pid.to_string());

        let mut pids: Vec<u32> = Vec::new();
        if self.cgroup_change_needed(get_cgroup(pid).as_deref()) {
            pids.push(pid);
        }
        for child in collect_child_pids(pid) {
            if self.cgroup_change_needed(get_cgroup(child).as_deref()) {
                pids.push(child);
            }
        }

        let sd_unit_c = sd_unit.clone();
        let sd_slice_c = sd_slice.clone();
        let dbus_conn = Arc::clone(&self.dbus_conn);

        let mut last_err: Option<Box<dyn std::error::Error + Send + Sync>> = None;
        for attempt in 1..=3u32 {
            let result =
                call_start_transient_unit(&dbus_conn, &sd_unit_c, &sd_slice_c, pids.clone()).await;

            match result {
                Ok(_) => {
                    debug!(
                        "window {} successfully assigned to cgroup {}/{}",
                        app_id, sd_slice, sd_unit
                    );
                    return Ok(());
                }
                Err(e) => {
                    warn!("assign_scope attempt {attempt}/3 failed: {e}");
                    last_err = Some(e);
                    tokio::time::sleep(Duration::from_millis(100 * attempt as u64)).await;
                }
            }
        }

        Err(last_err.unwrap())
    }

    async fn on_new_window(&self, event: WindowEvent) {
        let con = &event.container;
        let app_id = con.app_id.clone().or_else(|| {
            con.window_properties
                .as_ref()
                .and_then(|wp| wp.class.clone())
        });

        let pid = match con.pid {
            Some(p) if p > 0 => p as u32,
            _ => {
                warn!("Failed to get pid for {:?}", app_id);
                return;
            }
        };

        // Fallback: use process name if app_id is still unknown
        let app_id = app_id.unwrap_or_else(|| {
            fs::read_to_string(format!("/proc/{pid}/comm"))
                .unwrap_or_else(|_| pid.to_string())
                .trim()
                .to_string()
        });

        let cgroup = get_cgroup(pid);
        debug!("window {}({}) cgroup {:?}", app_id, pid, cgroup);

        if self.cgroup_change_needed(cgroup.as_deref()) {
            if let Err(e) = self.assign_scope(&app_id, pid).await {
                error!("Failed to modify cgroup for {}: {}", app_id, e);
            }
        }
    }
}

async fn call_start_transient_unit(
    conn: &Arc<Mutex<DbusConnection>>,
    unit: &str,
    slice: &str,
    pids: Vec<u32>,
) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
    let unit = unit.to_string();
    let slice = slice.to_string();
    let conn = Arc::clone(conn);

    tokio::task::spawn_blocking(move || {
        let conn = conn.blocking_lock();
        let proxy = conn.with_proxy(SD_BUS_NAME, SD_OBJECT_PATH, Duration::from_secs(15));

        let properties: Vec<(&str, Variant<Box<dyn RefArg>>)> = vec![
            ("PIDs", Variant(Box::new(pids) as Box<dyn RefArg>)),
            ("Slice", Variant(Box::new(slice) as Box<dyn RefArg>)),
        ];
        let aux: Vec<(&str, Vec<(&str, Variant<Box<dyn RefArg>>)>)> = vec![];

        let (_job,): (dbus::Path,) = proxy.method_call(
            SD_MANAGER_IFACE,
            "StartTransientUnit",
            (unit.as_str(), "fail", properties, aux),
        )?;
        Ok::<(), dbus::Error>(())
    })
    .await
    .map_err(|e| Box::new(e) as Box<dyn std::error::Error + Send + Sync>)?
    .map_err(|e| Box::new(e) as Box<dyn std::error::Error + Send + Sync>)
}

#[derive(Debug, Clone, ArgEnum)]
enum LogLevel {
    Critical,
    Error,
    Warning,
    Info,
    Debug,
}

impl LogLevel {
    fn to_filter(&self) -> log::LevelFilter {
        match self {
            LogLevel::Critical => log::LevelFilter::Error,
            LogLevel::Error => log::LevelFilter::Error,
            LogLevel::Warning => log::LevelFilter::Warn,
            LogLevel::Info => log::LevelFilter::Info,
            LogLevel::Debug => log::LevelFilter::Debug,
        }
    }
}

#[derive(Parser, Debug)]
#[clap(
    name = "assign-cgroups",
    about = "Assign CGroups to apps in compositors with i3/Sway IPC protocol support"
)]
struct Cli {
    #[clap(
        short,
        long,
        arg_enum,
        default_value = "info",
        help = "Set logging level"
    )]
    loglevel: LogLevel,
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let cli = Cli::parse();

    env_logger::builder()
        .filter_level(cli.loglevel.to_filter())
        .init();

    // D-Bus session connection
    let dbus_conn = Arc::new(Mutex::new(
        DbusConnection::new_session().map_err(|e| format!("DBus connection error: {e}"))?,
    ));

    // Locate Sway socket (for compositor PID via SO_PEERCRED)
    let socket_path = std::env::var("SWAYSOCK")
        .or_else(|_| std::env::var("I3SOCK"))
        .map_err(|_| "SWAYSOCK / I3SOCK not set — is Sway/i3 running?")?;

    // Compositor PID & cgroup
    let compositor_pid = get_pid_by_socket(&socket_path)?;
    let compositor_cgroup = get_cgroup(compositor_pid).ok_or("Could not read compositor cgroup")?;
    info!("compositor:{} {}", compositor_pid, compositor_cgroup);

    let handler = Arc::new(CGroupHandler::new(compositor_cgroup, dbus_conn));

    // Subscribe to window events via swayipc-async
    let mut events = SwayConnection::new()
        .await
        .map_err(|e| format!("Sway IPC connection error: {e}"))?
        .subscribe([EventType::Window])
        .await
        .map_err(|e| format!("Sway IPC subscribe error: {e}"))?;

    info!("Subscribed to window events on {}", socket_path);

    // Event loop
    while let Some(event) = events.next().await {
        let ev = match event {
            Ok(e) => e,
            Err(e) => {
                error!("Sway IPC error: {}", e);
                break;
            }
        };

        if let Event::Window(ev) = ev {
            if ev.change == WindowChange::New {
                let handler = Arc::clone(&handler);
                tokio::spawn(async move {
                    handler.on_new_window(*ev).await;
                });
            }
        }
    }

    Ok(())
}
